httpd/http_server.c

```diff
@@ -0,0 +1,333 @@
+}
+
+static int ssl_init()
+{
+    if (!config_get_boolean(HTTPD_CONFIG_SECTION, "ssl", false))
+        return 0;
+
+    char default_fn[FILENAME_MAX + 1];
+
+    snprintfz(default_fn,  FILENAME_MAX, "%s/ssl/key.pem",  netdata_configured_user_config_dir);
+    const char *key_fn  = config_get(HTTPD_CONFIG_SECTION, "ssl key", default_fn);
+
+    snprintfz(default_fn, FILENAME_MAX, "%s/ssl/cert.pem", netdata_configured_user_config_dir);
+    const char *cert_fn = config_get(HTTPD_CONFIG_SECTION, "ssl certificate",  default_fn);
+
+    accept_ctx.ssl_ctx = SSL_CTX_new(SSLv23_server_method());
+    SSL_CTX_set_options(accept_ctx.ssl_ctx, SSL_OP_NO_SSLv2);
+
+    /* load certificate and private key */
+    if (SSL_CTX_use_PrivateKey_file(accept_ctx.ssl_ctx, key_fn, SSL_FILETYPE_PEM) != 1) {
+        error("Could not load server key from \"%s\"", key_fn);
+        return -1;
+    }
+    if (SSL_CTX_use_certificate_file(accept_ctx.ssl_ctx, cert_fn, SSL_FILETYPE_PEM) != 1) {
+        error("Could not load certificate from \"%s\"", cert_fn);
+        return -1;
+    }
+
+    h2o_ssl_register_alpn_protocols(accept_ctx.ssl_ctx, h2o_http2_alpn_protocols);
+
+    return 0;
+}
+
+static h2o_pathconf_t *register_handler(h2o_hostconf_t *hostconf, const char *path, int (*on_req)(h2o_handler_t *, h2o_req_t *))
+{
+    h2o_pathconf_t *pathconf = h2o_config_register_path(hostconf, path, 0);
+    h2o_handler_t *handler = h2o_create_handler(pathconf, sizeof(*handler));
+    handler->on_req = on_req;
+    return pathconf;
+}
+
+// I did not find a way to do wildcard paths to make common handler for urls like:
+// /api/v1/info
+// /host/child/api/v1/info
+// /host/uuid/api/v1/info
+// ideally we could do something like "/*/api/v1/info" subscription
+// so we do it "manually" here with uberhandler
+static inline int _netdata_uberhandler(h2o_handler_t *self, h2o_req_t *req, RRDHOST **host)
+{
+    if (!h2o_memis(req->method.base, req->method.len, H2O_STRLIT("GET")))
+        return -1;
+
+    static h2o_generator_t generator = { NULL, NULL };
+
+    h2o_iovec_t norm_path = req->path_normalized;
+
+    if (norm_path.len > strlen(HOST_SELECT_PREFIX) && !memcmp(norm_path.base, HOST_SELECT_PREFIX, strlen(HOST_SELECT_PREFIX))) {
+        h2o_iovec_t host_id; // host_id can be either and UUID or a hostname of the child
+
+        norm_path.base += strlen(HOST_SELECT_PREFIX);
+        norm_path.len -= strlen(HOST_SELECT_PREFIX);
+
+        host_id = norm_path;
+
+        size_t end_loc = h2o_strstr(host_id.base, host_id.len, "/", 1);
+        if (end_loc != SIZE_MAX) {
+            host_id.len = end_loc;
+            norm_path.base += end_loc;
+            norm_path.len -= end_loc;
+        }
+
+        char *c_host_id = iovec_to_cstr(&host_id);
+        *host = rrdhost_find_by_hostname(c_host_id);
+        if (!*host)
+            *host = rrdhost_find_by_guid(c_host_id);
+        if (!*host) {
+            req->res.status = 400;
+            req->res.reason = "Wrong host id";
```

Comment:
Can this string on old compilers be allocated only in current stack? Should we not add it as a `define` to avoid this possible issue? :thinking:

Reply:
I don't think this is worth it. Only old compilers (what is our minimum considering C standard anyway) and even if so it is just few bytes.